pcdsdevices/interface.py

```diff
@@ -217,9 +218,162 @@ def __init_subclass__(cls, **kwargs):
+            Either the component objects themselves or strings that match the
+            attrs of the Components. If omitted, we'll go through all of the
+            Components.
+        include_lazy : bool, optional
+            If True, we'll also ensure the lazy components are ready.
+            Defaults to False, which means ignore the lazy components.
+        ensure_subs : bool, optional
+            If True, we'll instantiate the components that have subs, even if
+            they are lazy.
+            Defaults to True.
+        wait_connected : bool, optional
+            If True, we'll wait for everything to connect and raise a
+            TimeoutError if anything does not connect during the timeout
+            duration.
+            Defaults to False.
+        timeout : float, optional
+            The maximum time to wait for connections if wait_connected
+            is True. Defaults to 0.005s per component, or at least 1s.
+        """
+        if cpts is None:
+            activation_call = True
+            try:
+                cpts: Iterable[Component] = self._sig_attrs.values()
+            except AttributeError:
+                # No _sig_attrs means not a Device, meaning no cpts
+                self._active_flag = True
+                return
+            if wait_connected and timeout is None:
+                timeout = max(1.0, 0.005 * len(list(self.walk_components())))
+        else:
+            activation_call = False
+        if wait_connected and timeout is None:
+            timeout = max(1.0, 0.005 * len(cpts))
+        start = time.monotonic()
+        objs: List[Union[Device, Signal]] = []
+        for cpt in cpts:
+            if isinstance(cpt, str):
+                cpt: Component = self._sig_attrs[cpt]
+            if any((
+                cpt._subscriptions and ensure_subs,
+                include_lazy,
+                not cpt.lazy,
+            )):
+                objs.append(getattr(self, cpt.attr))
+        if activation_call:
+            self._active_flag = True
+        for obj in objs:
+            try:
+                obj.ensure_cpts()
+            except AttributeError:
+                pass
+            if not obj.connected and wait_connected:
+                elapsed = time.monotonic() - start
+                remaining = timeout - elapsed
+                if remaining <= 0:
+                    raise TimeoutError(
+                        'Timeout waiting for cpts to connect.'
+                    )
+                obj.wait_for_connection(timeout=remaining)
+
+    def destroy(self):
+        """
+        Disconnect and destroy all loaded signals and subdevices.
+
+        This is modified from Device.destroy() to consider that some
+        non-lazy signals may not have been created yet, else we risk
+        creating them during the destroy process.
+        """
+        self._destroyed = True
+        exceptions = []
+        for attr, obj in self._signals.items():
```

Comment:
I assume we don't use `walk_signals` because something implementing our `BaseInterface` might not be a `Device`, but if it isn't will it have `._signals`?

Reply:
The specific reason is that `walk_signals` will instantiate any components that the session has skipped, since it doesn't know what shenanigans I'm up to here, and I didn't want to create components on `destroy`. I also didn't want to change the `walk_signals` logic because I thought that the signal creation would theoretically be a desired behavior in this case.